File: spare-parts-shop/src/main/java/com/spareparts/model/EMI.java
package com.spareparts.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "emis")
@Data
@NoArgsConstructor
@AllArgsConstructor
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class EMI {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "business_id", nullable = false)
    @JsonIgnore
    private Business business;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "invoice_id", nullable = false)
    private Bill bill;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "customer_id", nullable = false)
    private Customer customer;

    @Column(name = "total_amount", nullable = false)
    private Double totalAmount;

    @Column(name = "down_payment")
    private Double downPayment = 0.0;

    @Column(name = "loan_amount", nullable = false)
    private Double loanAmount;

    @Column(name = "total_emis", nullable = false)
    private Integer totalEmis;

    @Column(name = "emi_amount", nullable = false)
    private Double emiAmount;

    @Column(name = "emis_paid")
    private Integer emisPaid = 0;

    @Column(name = "emis_remaining")
    private Integer emisRemaining;

    @Column(name = "paid_amount")
    private Double paidAmount = 0.0;

    @Column(name = "remaining_amount", nullable = false)
    private Double remainingAmount;

    @Column(name = "first_emi_date")
    private LocalDate firstEmiDate;

    @Column(name = "next_emi_date")
    private LocalDate nextEmiDate;

    @Column(name = "created_at")
    private LocalDateTime createdAt = LocalDateTime.now();
}

File: spare_parts_shop_flutter/lib/models/emi.dart

import 'customer.dart';
import 'bill.dart';

class EMI {
  final int id;
  final Bill bill;
  final Customer customer;
  final double totalAmount;
  final double downPayment;
  final double loanAmount;
  final int totalEmis;
  final double emiAmount;
  final int emisPaid;
  final int emisRemaining;
  final double paidAmount;
  final double remainingAmount;
  final double interestRate;
  final double processingFee;
  final String firstEmiDate;
  final String nextEmiDate;
  final String? emiNotes;
  final List<dynamic>? installments;
  final String createdAt;

  EMI({
    required this.id,
    required this.bill,
    required this.customer,
    required this.totalAmount,
    required this.downPayment,
    required this.loanAmount,
    required this.totalEmis,
    required this.emiAmount,
    required this.emisPaid,
    required this.emisRemaining,
    required this.paidAmount,
    required this.remainingAmount,
    required this.interestRate,
    required this.processingFee,
    required this.firstEmiDate,
    required this.nextEmiDate,
    this.emiNotes,
    this.installments,
    required this.createdAt,
  });

  factory EMI.fromJson(Map<String, dynamic> json) {
    return EMI(
      id: json['id'],
      bill: Bill.fromJson(json['bill']),
      customer: Customer.fromJson(json['customer']),
      totalAmount: (json['totalAmount'] as num).toDouble(),
      downPayment: (json['downPayment'] as num).toDouble(),
      loanAmount: (json['loanAmount'] as num).toDouble(),
      totalEmis: json['totalEmis'],
      emiAmount: (json['emiAmount'] as num).toDouble(),
      emisPaid: json['emisPaid'],
      emisRemaining: json['emisRemaining'],
      paidAmount: (json['paidAmount'] as num).toDouble(),
      remainingAmount: (json['remainingAmount'] as num).toDouble(),
      interestRate: (json['interestRate'] as num).toDouble(),
      processingFee: (json['processingFee'] as num).toDouble(),
      firstEmiDate: json['firstEmiDate'],
      nextEmiDate: json['nextEmiDate'],
      emiNotes: json['emiNotes'],
      installments: json['installments'],
      createdAt: json['createdAt'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'bill': bill.toJson(),
      'customer': customer.toJson(),
      'totalAmount': totalAmount,
      'downPayment': downPayment,
      'loanAmount': loanAmount,
      'totalEmis': totalEmis,
      'emiAmount': emiAmount,
      'emisPaid': emisPaid,
      'emisRemaining': emisRemaining,
      'paidAmount': paidAmount,
      'remainingAmount': remainingAmount,
      'interestRate': interestRate,
      'processingFee': processingFee,
      'firstEmiDate': firstEmiDate,
      'nextEmiDate': nextEmiDate,
      'emiNotes': emiNotes,
      'installments': installments,
      'createdAt': createdAt,
    };
  }
}
